### PR TITLE
explain should accept slug or uuid

### PR DESCRIPTION
**Pull request by kamaldodiya**
### Issue
Requesting explain endpoint 
- with slug works
    - https://staging-2020.ensembl.org/api/metadata/genome/grch38/explain
       ```
       curl http://localhost:8014/api/metadata/genome/grch38/explain | jq 
       {
       "genome_id": "a7335667-93e7-11ec-a39d-005056b38ce3",
       "genome_tag": "grch38",
        "common_name": "human",
        "scientific_name": "Homo sapiens",
        "type": null,
        "is_reference": true,
        "assembly": {
             "accession_id": "GCA_000001405.29",
             "name": "GRCh38.p14"
              }
       }
       ```
- with genome_uuid doesn't return anything
    - https://staging-2020.ensembl.org/api/metadata/genome/b29f62e6-992e-4049-a26b-b6bd673c51e7/explain
       ```
       curl http://localhost:8014/api/metadata/genome/a7335667-93e7-11ec-a39d-005056b38ce3/explain | jq 
        {}
       ```
  
### Solution

The explain endpoint should accept slug or genome_uuid. 

The fix will 
1) consider the input as slug
2) Try to find out genome_uuid for the slug
    a) if slug has genome_uuid then return the genome_uuid
    b) if slug has no genome_uuid consider input as genome_uuid
3) get genome details for the genome_uuid
4) prepare response and return

### Example
Request based on slug

```
curl http://localhost:8014/api/metadata/genome/grch38/explain | jq
{
  "genome_id": "a7335667-93e7-11ec-a39d-005056b38ce3",
  "genome_tag": "grch38",
  "common_name": "human",
  "scientific_name": "Homo sapiens",
  "type": null,
  "is_reference": true,
  "assembly": {
    "accession_id": "GCA_000001405.29",
    "name": "GRCh38.p14"
  }
}
```

Request based on genome_uuid

```
curl http://localhost:8014/api/metadata/genome/a7335667-93e7-11ec-a39d-005056b38ce3/explain | jq
{
  "genome_id": "a7335667-93e7-11ec-a39d-005056b38ce3",
  "genome_tag": "grch38",
  "common_name": "human",
  "scientific_name": "Homo sapiens",
  "type": null,
  "is_reference": true,
  "assembly": {
    "accession_id": "GCA_000001405.29",
    "name": "GRCh38.p14"
  }
}
```
If it doesn't exist
```
curl -sSD/dev/stderr 'http://localhost:8014/api/metadata/genome/just-kidding/explain' | jq
HTTP/1.1 404 Not Found
date: Tue, 31 Oct 2023 14:45:48 GMT
server: uvicorn
content-length: 53
content-type: application/json

{
  "message": "Could not  explain just-kidding"
}
``` 

for details endpoint
```
curl -sSD/dev/stderr 'http://localhost:8014/api/metadata/genome/just-kidding/details' | jq
HTTP/1.1 404 Not Found
date: Tue, 31 Oct 2023 16:11:40 GMT
server: uvicorn
content-length: 47
content-type: application/json

{
  "message": "Could not find details for just-kidding"
}
```

### CODE REVIEW

**Comment by azangru** [comment]
A general observation, based on the title of this PR: we should not think of the opposition as `slug` vs `genome_uuid`. The client sends to the api a string that it does not know anything about. In the api spec, I called this string a `slug`. This string can be either `genome_tag` or `genome_uuid`. Thus, the opposition is between `genome_tag` and `genome_uuid` rather than between the `slug` and `genome_uuid`.

I'll check how this is reflected in the code now.

File: app/api/resources/grpc_client.py [hunk 1]
```diff
@@ -85,4 +85,6 @@ def get_region(self, genome_uuid: str, region_name: str):
     def get_genome_uuid_from_slug(self, slug):
         uuid_request = ensembl_metadata_pb2.GenomeTagRequest(genome_tag=slug)
         genome_uuid_data = self.stub.GetGenomeUUIDByTag(uuid_request)
-        return genome_uuid_data.genome_uuid
\ No newline at end of file
+        if genome_uuid_data.genome_uuid:
+            return genome_uuid_data.genome_uuid
+        return slug
```
Comment (azangru) [inline app/api/resources/grpc_client.py:90]: You need to be able to handle garbage strings as well — you cannot assume that what the client sends you will always be a valid genome_tag or genome_uuid. Right now, this request:

https://staging-2020.ensembl.org/api/metadata/genome/just-kidding/details

will result in a beautiful 500 error:

image

Comment (kamaldodiya) [reply app/api/resources/grpc_client.py:90]: Yes, I just finished working on this I just need to push this.

Comment (kamaldodiya) [reply app/api/resources/grpc_client.py:90]: By the way endpoint in question here is explain not genome_details :)

Comment (kamaldodiya) [reply app/api/resources/grpc_client.py:90]:
```
curl 'http://localhost:8014/api/metadata/genome/just-kidding/explain' | jq
{
  "message": "Could not find details for just-kidding"
}
```

Comment (kamaldodiya) [reply app/api/resources/grpc_client.py:90]: @azangru  If you need specific message the do let us know and I will add it

Comment (azangru) [inline app/api/resources/grpc_client.py:90]: Also, the `get_genome_uuid_from_slug` is probably the wrong function to return the `slug` from. In pseudocode, I would imagine the following conversation with your code:

- Is this string I received from the client a genome_tag?
  - Yes, it it; cool; I know genome uuid now.
  - No, it isn't
- Hmm, I now either have a certain genome uuid resolved by genome_tag; or the original string from client that can be a genome uuid, or can be rubbish. Let's try and fetch genome info using the genome uuid that I've got.
  - Did I receive genome info from grpc? Great, return response according to spec
  - Did I not receive genome info? Then send back a 404 response.

Comment (kamaldodiya) [reply app/api/resources/grpc_client.py:90]: Yes, Just refresh the page :) I just implemented and pushed the changes :)

Comment (azangru) [reply app/api/resources/grpc_client.py:90]: The message is fine. The code won't read it though. I expect http status code will be a 404? 🙂 

For the future — when we start having releases, I expect both request and the response will become more complex (e.g. genome id may exist, but not for a particular release).

Comment (kamaldodiya) [reply app/api/resources/grpc_client.py:90]:
```
curl -sSD/dev/stderr 'http://localhost:8014/api/metadata/genome/just-kidding/explain' | jq
HTTP/1.1 404 Not Found
date: Tue, 31 Oct 2023 14:45:48 GMT
server: uvicorn
content-length: 53
content-type: application/json

{
  "message": "Could not find details for just-kidding"
}
```

Comment (azangru) [reply app/api/resources/grpc_client.py:90]: It's just that I would expect the function called `get_genome_uuid_from_slug` to work differently — to return a real genome uuid the string is a genome tag, or None if it isn't (and also, probably to be called `get_genome_uuid_from_genome_tag` for clarity). As it is written, it returns either a real genome uuid or a garbage string.

But it is up to you to choose which is better.

File: app/api/resources/metadata.py [hunk 2]
```diff
@@ -104,16 +104,37 @@ def example_objects(request: Request, genome_id: str):
...
+    uuid_from_tag = grpc_client.get_genome_uuid_from_tag(slug)
     response_dict = {}
-    if genome_uuid:
-        genome_details_dict = MessageToDict(grpc_client.get_genome_details(genome_uuid))
-        genome_details = GenomeDetails(**genome_details_dict)
-        response_dict = genome_details.model_dump(include={"genome_id":True, "genome_tag":True, "scientific_name":True, "common_name":True, "is_reference" : True, "assembly": {"name", "accession_id"}, "type": True})
-    return responses.JSONResponse(response_dict)
+    response_data = None
+    if uuid_from_tag:
+        genome_detail_for_genome = uuid_from_tag
```
Comment (azangru) [inline app/api/resources/metadata.py:126]: `genome_detail_for_genome` is just a `genome_uuid`, isn't it? A string that we hope (but not perfectly sure; but that's all right) is a genome uuid.

Would it be easier if line 122 said `genome_uuid = grpc_client.get_genome_uuid_from_slug(slug) or slug`?

**Review by azangru — APPROVED** [review]

**Review by Mehrnaz-Charkhchi — APPROVED** [review]

File: app/api/models/genome.py [hunk 3]
```diff
@@ -51,8 +51,8 @@ class GenomeDetails(BaseModel):
     assembly_level: str = Field(alias=AliasPath('attributesInfo', 'assemblyLevel'))
     assembly_date: str = Field(alias=AliasPath('attributesInfo', 'assemblyDate'), default=None)
     annotation_provider: AnnotationProvider = None
-    annotation_method: str = Field(alias='genebuildMethod', default=None)
```
Comment (Mehrnaz-Charkhchi) [inline app/api/models/genome.py:54]: @azangru @kamaldodiya I added the mappings to this PR because these are small updates.

**Pull request merged**